### PR TITLE
19265 added Continuation Authorization component

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#19265

*Description of changes:*
- app version = 5.10.5
- added hint prop to file upload component
- misc renaming in file upload component
- misc cleanup in file upload component (not a refactor)
- added Continuation Authorization component
- updated continuation authorization code in summary component
- very minor cleanup of Complete Affidait component (re: file upload)
- very minor cleanup of Upload Memorandum component
- very minor cleanup of Upload Rules component
- added Continuation Authorization interface
- added misc comments
- updated filing build/parse code
- updated Cont In first step

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 1]
```diff
@@ -0,0 +1,393 @@
+<template>
+  <!--
+    Don't render this component until authorization object is initialized, so that we mount
+    FileUploadPreview with existing files, because its "inputFile" prop is not reactive.
+  -->
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:5]: Ketaki, this is one trick to using the FileUploadPreview component. (`inputFile` is only accepted when the component is rendered/mounted)

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-682-jnuvr89x.web.app

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+          @fileSelected="onFileSelected(index, $event)"
+        />
+      </v-col>
+    </v-row>
+
+    <v-form
+      ref="formRef"
+      lazy-validation
+      @submit.prevent
+    >
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:62]: A `v-form` is needed around the native Vuetify components to trigger them to validate.

Note that the FileUploadPreview and DatePickerShared components need to be validated independently -- see line 325 below.

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+          sm="9"
+          class="pl-8"
+        >
+          <v-text-field
+            v-model="authorization.authorityName"
+            class="authority-name"
+            filled
+            hide-details="auto"
+            label="Authority Name"
+            :rules="getShowErrors ? authorityNameRules : []"
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:85]: This is a small trick to enable the rules only once the global "show errors" flag is set to True.

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+          />
+        </v-col>
+      </v-row>
+
+      <v-checkbox
+        id="continuation-authorization-checkbox"
+        v-model="authorization.isConfirmed"
+        class="mt-6"
+        hide-details
+        :rules="getShowErrors ? [(v) => !!v] : []"
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:153]: Easy way to make this a "required" field.

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+  }
+
+  /** The number of file upload componenets to display (max 5). */
+  get numUploads (): number {
+    return Math.min((this.authorization.files.length + 1), 5)
+  }
+
+  mounted (): void {
+    // set or initialize authorization object
+    this.authorization = this.getContinuationAuthorization || { files: [] } as ContinuationAuthorizationIF
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:255]: Note: until this point, the `authorization` object was null. Now it's either whatever is in the store (resumed from a draft) or an almost-empty object (which is fine).

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    this.fileValidity = valid
+  }
+
+  /**
+   * Called when FileUploadPreview tells us about a new or cleared file.
+   * This is called right after the File Validity event.
+   */
+  async onFileSelected (index: number, file: File): Promise<void> {
+    // reset state of file uploader to ensure not in manual error mode
+    this.fileUploadCustomErrorMsg = ''
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:272]: See also line 297.

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+  mounted (): void {
+    // set or initialize authorization object
+    this.authorization = this.getContinuationAuthorization || { files: [] } as ContinuationAuthorizationIF
+  }
+
+  /**
+   * Called when FileUploadPreview tells us whether a file is valid.
+   * This is called right before the File Selected event.
+   */
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:261]: This is another trick to using the file upload component: knowing that the file-validity event comes right before the file-selected event.

Since there's only 1 file uploaded at a time, I simply set a variable (line 263) and then check it (line 276). No need for an array.

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+        this.snackbar = true
+        return
+      }
+
+      // try to upload to Minio
+      let psu: PresignedUrlIF
+      try {
+        psu = await this.getPresignedUrl(file.name)
+        const res = await this.uploadToUrl(psu.preSignedUrl, file, psu.key, this.getKeycloakGuid)
+        if (!res || res.status !== 200) throw new Error()
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:294]: I compressed the above from a separate method (as in other filings) to these few lines. That's all that's needed to do the actual upload!

File: src/components/ContinuationIn/ContinuationAuthorization.vue [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+  private onShowErrors (): void {
+    if (this.getShowErrors) {
+      // validate the form and the components that don't support form validation
+      this.$refs.formRef.validate()
+      this.$refs.authorizationDateRef.validateForm()
+      this.$refs.expiryDateRef.validateForm()
+    }
+  }
+
+  @Watch('authorization', { deep: true })
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ContinuationAuthorization.vue:332]: Whenever the authorization object reference changes, or any of its properties, this method gets called.

File: src/components/ContinuationIn/ExtraproRegistration.vue [hunk 10]
```diff
@@ -384,7 +382,11 @@ export default class ExtraproRegistration extends Mixins(DocumentMixin) {
   }
 
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  async fileSelected (file): Promise<void> {
+  onFileValidity (valid: boolean): void {
+    // *** TODO: implement -- see UploadMemorandum.vue
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/ExtraproRegistration.vue:386]: This will be done in ticket entity/bcgov#21138.

File: src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue [hunk 11]
```diff
@@ -170,13 +170,45 @@
...
-              </div>
+              <ul id="continuation-authorization-file">
+                <li
+                  v-for="(item, index) in getContinuationAuthorization?.files"
+                  :key="index"
+                >
+                  <v-icon color="green darken-2">
+                    mdi-check
+                  </v-icon>
+                  <span>{{ getContinuationAuthorization?.files[index]?.fileName }}</span>
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue:181]: I think I can simplify this to `item.fileName` ;)

File: src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue [hunk 12]
```diff
@@ -215,7 +247,8 @@
               class="pt-4 pt-sm-0"
             >
               <div id="authorization-date">
-                {{ authorizationDate || '[Unknown]' }}
+                {{ yyyyMmDdToPacificDate(getContinuationAuthorization?.date, true, false) ||
+                  '[Unknown]' }}
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue:251]: Do you think I should move this into a computed (getter)?

File: src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue [hunk 13]
```diff
@@ -355,4 +375,18 @@ article:not(:first-child) {
...
+    list-style: none;
+    margin-left: 2rem;
+    padding-left: 0;
+
+    li > i {
+      margin-left: -2rem;
+      padding-right: 10px;
+    }
+  }
+}
```
Comment (severinbeauvais) [inline src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue:391]: This is to line the icon with the column, have proper whitespace, and indent the text block, ie:

image

File: src/components/common/FileUploadPreview.vue [hunk 14]
```diff
@@ -47,14 +50,14 @@ export default class FileUploadPreview extends Mixins(DocumentMixin) {
...
+      (file: File) => {
         if (this.isRequired) {
           return !!file || this.inputFileLabel + ' is required'
         }
         return true
       },
-      (file) => {
-        if (this.maxSize) {
+      (file: File) => {
+        if (file && this.maxSize) {
```
Comment (severinbeauvais) [inline src/components/common/FileUploadPreview.vue:60]: Bug fix: ensure file exists before computing this rule!

File: src/components/common/FileUploadPreview.vue [hunk 15]
```diff
@@ -68,23 +71,20 @@ export default class FileUploadPreview extends Mixins(DocumentMixin) {
...
     if (this.inputFile) {
       this.fileUpload = this.inputFile
+      // wait for v-file-input to settle
       await this.$nextTick()
-      this.isFileValid(this.fileUpload)
     }
-  }
-
-  async fileChange (file): Promise<void> {
-    await this.emitFileSelected(file)
```
Comment (severinbeauvais) [inline src/components/common/FileUploadPreview.vue:77]: This is was only called by the template so I combined it into emitFileSelected() (and then renamed it). This is minor cleanup, not a refactor.

File: src/components/common/FileUploadPreview.vue [hunk 16]
```diff
@@ -68,23 +71,20 @@ export default class FileUploadPreview extends Mixins(DocumentMixin) {
...
   // Note: the validation is done this way as opposed to being all defined in the validation rules(fileUploadRules)
   //  above because Vuetify does not support async validation out of the box.  This was needed to work
   //  around this limitation.  vuelidate does support this but this would mean introducing a component that
   //  is using a different validation approach or updating all components to use vuelidate.  Have decided
   //  to do this for the time being.
-  private async validateFileInput (file): Promise<boolean> {
+  private async validateFileInput (file: File): Promise<boolean> {
     this.customErrorMessages = []
-    const isValid = this.$refs.fileUploadInput.validate() as boolean
+    const isValid = this.$refs.form.validate()
```
Comment (severinbeauvais) [inline src/components/common/FileUploadPreview.vue:87]: As on the other page, the Vuetify v-file-input component needs to be wrapped in a form to trigger its validation. (Otherwise, this change is just a variable rename.)

File: src/interfaces/utils-interfaces/presigned-url-interface.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+/** Response object from LegalServices.getPresignedUrl(). */
```
Comment (severinbeauvais) [inline src/interfaces/utils-interfaces/presigned-url-interface.ts:1]: This was previously named "document-upload-interface".

File: src/mixins/document-mixin.ts [hunk 18]
```diff
@@ -37,12 +42,17 @@ export default class DocumentMixin extends Vue {
           throw new Error('Invalid API response')
         }
         return data
-      }).catch(error => {
-        console.log('Error getting presigned url =', error) // eslint-disable-line no-console
-        throw error
```
Comment (severinbeauvais) [inline src/mixins/document-mixin.ts:42]: This is redundant. The Axios error was already thrown and displayed a console message..

File: src/mixins/filing-template-mixin.ts [hunk 19]
```diff
@@ -452,14 +447,11 @@ export default class FilingTemplateMixin extends Mixins(AmalgamationMixin, DateM
...
-      continuationIn.business &&
-      continuationIn.foreignJurisdiction
-    ) {
+    if (continuationIn.foreignJurisdiction) {
       this.setExistingBusinessInfo({
         affidavitFileKey: continuationIn.foreignJurisdiction.affidavitFileKey,
-        businessIdentifier: continuationIn.business.identifier,
-        businessLegalName: continuationIn.business.legalName,
+        businessIdentifier: continuationIn.business?.identifier, // FUTURE: need fallback to null?
+        businessLegalName: continuationIn.business?.legalName, // FUTURE: need fallback to null?
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:454]: Ketaki, I had second thoughts about this... If there was no business object, should these fields be null or undefined?

Probably need to test without an expro (ie, your ticket) and see what makes the most sense. For example, does everything still work if the user clicks Undo on the manual component and then searches for an expro?

File: src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue (hunk 11 above)
Comment (severinbeauvais) [reply src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue]: Done.

File: src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue (hunk 12 above)
Comment (ketaki-deodhar) [reply src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue:251]: might be a good idea to move but okay to keep like this as well

Comment (severinbeauvais) [reply src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue]: I moved it.

Comment (severinbeauvais) [reply src/components/ContinuationIn/SummaryBusinessHomeJurisdiction.vue]: I'm all for good ideas, so I moved it.

File: src/mixins/filing-template-mixin.ts (hunk 19 above)
Comment (ketaki-deodhar) [reply src/mixins/filing-template-mixin.ts:454]: Okay I will keep a note of this and test it

**Review by ketaki-deodhar — APPROVED** [review]
Tested with some uploads and looks really good!

**Review by vysakh-menon-aot — APPROVED** [review]

**Comment by severinbeauvais** [comment]
This code change is functional and demoable, but I'm still working on an edge case and unit tests (future PR).
